### DISCUSSION TITLE
Performance comparison between duckdb and parquet.

### DISCUSSION THREAD

**Discussion by neverchanje** [duckdb/duckdb#5638] [Show and tell]
Hi DuckDB folks, I just wanna show you some benchmark results which I found pretty interesting. The entire benchmark consists of 4 cases. I used the testing data downloaded from:
- https://github.com/duckdb/duckdb-data/releases/download/v1.0/taxi_2019_04.parquet
- https://github.com/duckdb/duckdb-data/releases/download/v1.0/taxi_2019_05.parquet

I performed all benchmarks on my Macbook Pro 2021, Apple M1 Pro, 32 GB memory.

## Test Case 1

```sql
INSTALL parquet; LOAD parquet;

-- View, 14998399 rows
CREATE VIEW taxi AS
SELECT * FROM read_parquet('./taxi_2019_04.parquet')
UNION ALL
SELECT * FROM read_parquet('./taxi_2019_05.parquet');

.timer on

-- Test Query
select passenger_count, count(*) from taxi group by passenger_count;
```

UNION: <del>Run Time (s): real 1.038 user 7.344208 sys 0.855206</del>
UNION ALL: Run Time (s): real 0.036 user 0.231862 sys 0.034088

## Test Case 2

```sql
INSTALL parquet; LOAD parquet;

-- View, 14998400 rows.
CREATE VIEW taxi AS
SELECT * FROM read_parquet(['./taxi_2019_04.parquet', './taxi_2019_05.parquet']);

.timer on

-- Test Query
select passenger_count, count(*) from taxi group by passenger_count;
```

Run Time (s): real 0.037 user 0.239891 sys 0.028778

From test case 1 and 2, we can conclude that the best practice of reading from external parquet files is to pass a list of files to `read_parquet([])` instead of using `UNION`. `UNION ALL` can achieve similar performance with `read_parquet([])`.

## Test Case 3

```sql
INSTALL parquet; LOAD parquet;

-- Table
CREATE TABLE taxi AS
SELECT * FROM read_parquet(['./taxi_2019_04.parquet', './taxi_2019_05.parquet']);

.timer on

-- Test Query
select passenger_count, count(*) from taxi group by passenger_count;
```

Run Time (s): real 0.020 user 0.163598 sys 0.001290

- taxi_2019_04.parquet: 121MB
- taxi_2019_05.parquet: 124MB
- taxi_2019_05.db: 469MB

By materializing parquet files into DuckDB, the performance increased 2x (0.37/0.20). The result might not be accurate but I guess the reason is because DuckDB's compression policy is less aggresive than Parquet, as its file size is also 2x larger.

## Test Case 4

```sql
INSTALL parquet; LOAD parquet;

CREATE TABLE taxi AS
SELECT * FROM read_parquet(['./taxi_2019_04.parquet', './taxi_2019_05.parquet']);

COPY (SELECT * FROM taxi) TO 'taxi.parquet' (FORMAT 'parquet');
```

- taxi_2019_04.parquet: 121MB
- taxi_2019_05.parquet: 124MB
- taxi.parquet: 379MB

```sql
INSTALL parquet; LOAD parquet;

CREATE VIEW taxi AS
SELECT * FROM read_parquet(['./taxi.parquet']);

.timer on

-- Test Query
select passenger_count, count(*) from taxi group by passenger_count;
```

Run Time (s): real 0.036 user 0.255940 sys 0.007726

When I imported the two files and merged them into one single file, the file size became 1.54x larger. However, the query performance on the consolidated Parquet file was still the same with which on two Parquet files.

# Conclusion

I'm not an expert to DuckDB's execution engine, TBH. But I think we can still learn from the above benchmarking that if we want to support reading from external DuckDB, we should take the performance of reading multiple DuckDB files into consideration. <del>The interface should be like `read_duckdb([('./test_1.db', 'test_1_table'), ('<db file>', '<table_name>')])`. Or maybe it can be provided as a overloaded function.</del>

See also: https://github.com/duckdb/duckdb/pull/5048

**Comment by Alex-Monahan** [comment]
Could you re-run using UNION ALL? Union has a de-duplication step that makes this an imperfect benchmark.

**Comment by neverchanje** [comment]
Thanks Alex, UNION ALL did achieve a similar performance with 'read_parquet([])'. 😄

**Comment by l1t1** [comment]
try https://duckdb.org/docs/data/parquet
-- read all files that match the glob pattern
SELECT * FROM 'test/*.parquet';
